src/Features/CSharp/Portable/Snippets/ConsoleSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+
+        protected override async Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return false;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var syntaxContext = (CSharpSyntaxContext)document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
+
+            if (!ShouldDisplaySnippet(syntaxContext))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool ShouldDisplaySnippet(CSharpSyntaxContext context)
+        {
+            var token = context.LeftToken;
+            var isDirectlyInUndesirableLocation = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.ParameterList) ||
+                             node.IsKind(SyntaxKind.SimpleLambdaExpression) ||
+                             node.IsKind(SyntaxKind.ArgumentList) ||
+                             node.IsKind(SyntaxKind.RecordDeclaration) ||
+                             node.IsKind(SyntaxKind.ObjectCreationExpression) ||
+                             node.IsKind(SyntaxKind.SwitchExpression));
+
+            if (isDirectlyInUndesirableLocation)
+            {
+                return false;
+            }
+
+            var isExpressionInVariable = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.ParenthesizedLambdaExpression) || node.IsKind(SyntaxKind.AnonymousMethodExpression));
+            var isInVariableDeclaration = token.GetAncestors<SyntaxNode>().Any(node => node.IsKind(SyntaxKind.VariableDeclaration));
+
+            if (isInVariableDeclaration && !isExpressionInVariable)
+            {
+                return false;
+            }
+
+            var isInsideMethod = token.GetAncestors<SyntaxNode>()
+               .Any(node => node.IsKind(SyntaxKind.MethodDeclaration) ||
+                            node.IsKind(SyntaxKind.ConstructorDeclaration) ||
+                            node.IsKind(SyntaxKind.LocalFunctionStatement) ||
+                            node.IsKind(SyntaxKind.AnonymousMethodExpression) ||
+                            node.IsKind(SyntaxKind.ParenthesizedLambdaExpression));
+
+            var isInNamespace = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.NamespaceDeclaration) ||
+                             node.IsKind(SyntaxKind.FileScopedNamespaceDeclaration));
+
+            if (isInNamespace && !isInsideMethod)
```

Comment:
This check seems redundant.. the one below covers all cases doesn't it?

Reply:
```
namespace Namespace;
$$
class Program
{
    public async Task MethodAsync()
    {
    }
}
```

In that case, your cursor position is located in a global context but not a method.